File: src/main/java/me/alegian/thavma/impl/common/data/capability/AspectContainer.java
package me.alegian.thavma.impl.common.data.capability;

import me.alegian.thavma.impl.common.aspect.Aspect;
import me.alegian.thavma.impl.common.aspect.AspectMap;
import me.alegian.thavma.impl.common.aspect.AspectStack;
import me.alegian.thavma.impl.init.registries.T7Capabilities;
import me.alegian.thavma.impl.init.registries.deferred.Aspects;
import me.alegian.thavma.impl.init.registries.deferred.T7DataComponents;
import net.minecraft.core.BlockPos;
import net.minecraft.world.InteractionHand;
import net.minecraft.world.entity.player.Player;
import net.minecraft.world.item.ItemStack;
import net.minecraft.world.level.Level;
import net.minecraft.world.level.block.entity.BlockEntity;
import net.neoforged.neoforge.common.MutableDataComponentHolder;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import java.util.Optional;

public class AspectContainer implements IAspectContainer {
  private final MutableDataComponentHolder holder;
  private final int capacity;
  private final boolean visSource;
  private final boolean essentiaSource;

  public AspectContainer(MutableDataComponentHolder holder, int capacity, boolean visSource, boolean essentiaSource) {
    this.holder = holder;
    this.capacity = capacity;
    this.visSource = visSource;
    this.essentiaSource = essentiaSource;
  }

  public AspectContainer(MutableDataComponentHolder holder, int capacity) {
    this(holder, capacity, false, false);
  }

  public AspectContainer(MutableDataComponentHolder holder) {
    this(holder, Integer.MAX_VALUE);
  }

  public static Optional<IAspectContainer> at(Level level, BlockPos pos) {
    return Optional.ofNullable(level.getCapability(T7Capabilities.AspectContainer.INSTANCE.getBLOCK(), pos, null));
  }

  public static IAspectContainer getAspectContainerInHand(Player player) {
    var mainHandItem = player.getItemInHand(InteractionHand.MAIN_HAND);
    var offHandItem = player.getItemInHand(InteractionHand.OFF_HAND);

    IAspectContainer aspectContainer = null;

    if (!mainHandItem.isEmpty())
      aspectContainer = mainHandItem.getCapability(T7Capabilities.AspectContainer.INSTANCE.getITEM());
    else if (!offHandItem.isEmpty())
      aspectContainer = offHandItem.getCapability(T7Capabilities.AspectContainer.INSTANCE.getITEM());

    return aspectContainer;
  }

  public static Optional<IAspectContainer> from(ItemStack itemStack) {
    return Optional.ofNullable(itemStack.getCapability(T7Capabilities.AspectContainer.INSTANCE.getITEM()));
  }

  public static Optional<IAspectContainer> from(BlockEntity be) {
    return Optional.ofNullable(be.getLevel()).map(l -> l.getCapability(T7Capabilities.AspectContainer.INSTANCE.getBLOCK(), be.getBlockPos()));
  }

  public static boolean isAspectContainer(Level level, BlockPos blockPos) {
    return level.getCapability(T7Capabilities.AspectContainer.INSTANCE.getBLOCK(), blockPos) != null;
  }

  public static Optional<Pair> blockSourceItemSink(Level level, BlockPos blockPos, ItemStack itemStack) {
    return AspectContainer.from(itemStack).flatMap(sink ->
        AspectContainer.at(level, blockPos).map(source -> new Pair(source, sink))
    );
  }

  @Override
  public @NotNull AspectMap getAspects() {
    AspectMap aspectMap = this.holder.get(T7DataComponents.INSTANCE.getASPECTS());
    if (aspectMap == null) return AspectMap.EMPTY;
    return aspectMap;
  }

  @Override
  public void setAspects(AspectMap aspects) {
    this.holder.set(T7DataComponents.INSTANCE.getASPECTS(), aspects);
  }

  @Override
  public boolean areAspectsNull() {
    return this.holder.get(T7DataComponents.INSTANCE.getASPECTS()) == null;
  }

  @Override
  public int insert(Aspect aspect, int amount, boolean simulate) {
    if (amount == 0) return 0;

    AspectMap current = this.getAspects();
    var maxInsert = this.getCapacity() - current.get(aspect);
    var cappedInsert = Math.min(amount, maxInsert);

    if (!simulate) this.setAspects(current.add(aspect, cappedInsert));

    return cappedInsert;
  }

  @Override
  public int extract(Aspect aspect, int amount, boolean simulate) {
    if (amount == 0) return 0;
    var maxSubtract = this.getAspects().get(aspect);
    var cappedSubtract = Math.min(amount, maxSubtract);

    if (!simulate) this.setAspects(this.getAspects().subtract(aspect, amount));

    return cappedSubtract;
  }

  @Override
  public int getCapacity() {
    return this.capacity;
  }

  @Override
  public boolean isVisSource() {
    return this.visSource;
  }

  @Override
  public boolean isEssentiaSource() {
    return this.essentiaSource;
  }

  public static class Pair {
    private final IAspectContainer source;
    private final IAspectContainer sink;

    public Pair(IAspectContainer source, IAspectContainer sink) {
      this.source = source;
      this.sink = sink;
    }

    protected int simulateTransfer(Aspect a, int idealAmount) {
      int maxInsert = this.sink.insert(a, idealAmount, true);
      return this.source.extract(a, maxInsert, true);
    }

    public boolean canTransferPrimals() {
      return Aspects.INSTANCE.getPRIMAL_ASPECTS().stream()
          .map(a -> this.simulateTransfer(a.get(), 1))
          .anyMatch(e -> e > 0);
    }

    public @Nullable AspectStack transferPrimal(int indexOffset, int idealAmount) {
      var primals = Aspects.INSTANCE.getPRIMAL_ASPECTS().size();
      for (int i = 0; i < primals; i++) {
        var a = Aspects.INSTANCE.getPRIMAL_ASPECTS().get((i + indexOffset) % primals).get();
        int amount = this.simulateTransfer(a, idealAmount);
        if (amount == 0) continue;
        this.sink.insert(a, amount, false);
        this.source.extract(a, amount, false);
        return AspectStack.Companion.of(a, amount);
      }
      return null;
    }
  }
}

File: src/main/java/me/alegian/thavma/impl/common/data/capability/AspectContainer.kt
package me.alegian.thavma.impl.common.data.capability

import me.alegian.thavma.impl.common.aspect.Aspect
import me.alegian.thavma.impl.common.aspect.AspectMap
import me.alegian.thavma.impl.common.aspect.AspectStack
import me.alegian.thavma.impl.common.aspect.AspectStack.Companion.of
import me.alegian.thavma.impl.init.registries.T7Capabilities.AspectContainer.BLOCK
import me.alegian.thavma.impl.init.registries.T7Capabilities.AspectContainer.ITEM
import me.alegian.thavma.impl.init.registries.deferred.Aspects.PRIMAL_ASPECTS
import me.alegian.thavma.impl.init.registries.deferred.T7DataComponents.ASPECTS
import net.minecraft.core.BlockPos
import net.minecraft.world.item.ItemStack
import net.minecraft.world.level.Level
import net.minecraft.world.level.block.entity.BlockEntity
import net.neoforged.neoforge.common.MutableDataComponentHolder
import kotlin.math.min

open class AspectContainer(
  private val holder: MutableDataComponentHolder,
  override val capacity: Int = Int.MAX_VALUE,
  val slots: Int = Int.MAX_VALUE
) : IAspectContainer {
  override var aspects = holder.get(ASPECTS) ?: AspectMap()
    set(a) {
      holder.set(ASPECTS, a)
    }

  override fun areAspectsNull(): Boolean {
    return holder.get(ASPECTS) == null
  }

  override fun insert(aspect: Aspect, amount: Int, simulate: Boolean): Int {
    if (amount == 0) return 0
    if (aspects.size >= slots && !aspects.has(aspect)) return 0

    val current = this.aspects
    val maxInsert = this.capacity - current[aspect]
    val cappedInsert = min(amount.toDouble(), maxInsert.toDouble()).toInt()

    if (!simulate) this.aspects = current.add(aspect, cappedInsert)

    return cappedInsert
  }

  override fun extract(aspect: Aspect, amount: Int, simulate: Boolean): Int {
    if (amount == 0) return 0
    val maxSubtract = this.aspects[aspect]
    val cappedSubtract = min(amount.toDouble(), maxSubtract.toDouble()).toInt()

    if (!simulate) this.aspects = this.aspects.subtract(aspect, amount)

    return cappedSubtract
  }

  open class Pair(private val source: IAspectContainer, private val sink: IAspectContainer) {
    private fun simulateTransfer(a: Aspect, idealAmount: Int): Int {
      val maxInsert = sink.insert(a, idealAmount, true)
      return source.extract(a, maxInsert, true)
    }

    open fun canTransferPrimals(): Boolean {
      return PRIMAL_ASPECTS.stream()
        .anyMatch { a -> simulateTransfer(a.get(), 1) > 0 }
    }

    open fun transferAnyAspect(): AspectStack? {
      for (a in source.aspects.map { it.aspect }) {
        val amount = simulateTransfer(a, source.aspects[a])
        if (amount == 0) continue
        sink.insert(a, amount, false)
        source.extract(a, amount, false)
        return of(a, amount)
      }
      return null
    }

    open fun transferAllAspects() {
      for (a in source.aspects.map { it.aspect }) {
        val amount = simulateTransfer(a, source.aspects[a])
        if (amount == 0) continue
        sink.insert(a, amount, false)
        source.extract(a, amount, false)
      }
    }

    open fun transferPrimal(indexOffset: Int, idealAmount: Int): AspectStack? {
      val primals = PRIMAL_ASPECTS.size
      for (i in 0..<primals) {
        val a = PRIMAL_ASPECTS[(i + indexOffset) % primals].get()
        val amount = simulateTransfer(a, idealAmount)
        if (amount == 0) continue
        sink.insert(a, amount, false)
        source.extract(a, amount, false)
        return of(a, amount)
      }
      return null
    }
  }

  companion object {
    fun at(level: Level?, pos: BlockPos): IAspectContainer? {
      return level?.getCapability(BLOCK, pos)
    }

    fun from(itemStack: ItemStack): IAspectContainer? {
      return itemStack.getCapability(ITEM)
    }

    fun from(be: BlockEntity): IAspectContainer? {
      return be.level?.getCapability(BLOCK, be.blockPos, null, be)
    }

    fun isAspectContainer(level: Level, blockPos: BlockPos): Boolean {
      return level.getCapability(BLOCK, blockPos) != null
    }

    fun blockSourceItemSink(level: Level, blockPos: BlockPos, itemStack: ItemStack): Pair? {
      val sink = from(itemStack)
      val source = at(level, blockPos)
      if (sink == null || source == null) return null
      return Pair(source, sink)
    }

    fun itemSourceBlockSink(level: Level, blockPos: BlockPos, itemStack: ItemStack): Pair? {
      val sink = at(level, blockPos)
      val source = from(itemStack)
      if (sink == null || source == null) return null
      return Pair(source, sink)
    }
  }
}
